app/recurring_job.go

```diff
@@ -375,7 +422,48 @@ func (job *Job) doSnapshot() (err error) {
 	return nil
 }
 
-func (job *Job) doSnapshotCleanup(backupDone bool) (err error) {
+func (job *Job) getIsSnapshotDeleteFirst() (bool, error) {
+	SnapshotDeleteFirstSetting, err := job.lhClient.LonghornV1beta2().Settings(job.namespace).Get(context.TODO(), string(types.SettingNameRecurringJoBSnapshotDeleteFirst), metav1.GetOptions{})
+	if err != nil {
+		job.logger.WithError(err).Errorf("Failed to get %v setting", types.SettingNameRecurringJoBSnapshotDeleteFirst)
+		return false, err
+	}
+
+	isSnapshotDeleteFirst, err := strconv.ParseBool(SnapshotDeleteFirstSetting.Value)
+	if err != nil {
+		job.logger.WithError(err).Errorf("Failed to parse boolean value for setting %v", types.SettingNameRecurringJoBSnapshotDeleteFirst)
+		return false, err
+	}
+
+	return isSnapshotDeleteFirst, nil
+}
+
+func (job *Job) deleteSnapshotRetainReachedOrExceeded(volume *longhornclient.Volume, isSnapshotDeleteFirst bool) error {
+	volumeAPI := job.api.Volume
+
+	collection, err := volumeAPI.ActionSnapshotList(volume)
+	if err != nil {
+		return err
+	}
+
+	jobSnapshots := filterSnapshotsWithLabel(collection.Data, types.RecurringJobLabel, job.labels[types.RecurringJobLabel])
+	if len(jobSnapshots) >= job.retain {
```

Comment:
Should this one be strictly bigger:  len(jobSnapshots) > job.retain so that we don't have case in which number snapshot < expected retaining count after the deletion?

Reply:
As `MaxRecurringJobRetain = 50` in AD controller?

Reply:
No, I mean if we delete the snapshot when len(jobSnapshots) == job.retain and later on if we fail to create a new snapshot, the number of snapshot will be less than expected, right? 

If that cycle repeats, we will have less and less snapshot. So, I am wondering if the condition should be len(jobSnapshots) > job.retain

Reply:
Doin this would fix the issue https://github.com/longhorn/longhorn-manager/pull/1708#issuecomment-1441260982 too

Reply:
In `listSnapshotNamesToCleanup` of `doSnapshotCleanup` to [filterExpiredItems(snapshotsToNameWithTimestamps(snapshots), job.retain)](https://github.com/longhorn/longhorn-manager/pull/1708/files#diff-61e6464dd7934db9eb985d4107eb81d205c39647995e386a44401980ffbfa175R574-R579)
IIUC the [filterExpiredItems](https://github.com/longhorn/longhorn-manager/blob/master/app/recurring_job.go#L791-L801) only list the number of old snapshots that exceeds the `job.retain` or `job.retain-1` in this PR.